src/jl_uv.c

```diff
@@ -678,56 +678,71 @@ JL_DLLEXPORT int jl_printf(uv_stream_t *s, const char *format, ...)
     return c;
 }
 
-STATIC_INLINE void print_error_msg_as_json(char *buf) JL_NOTSAFEPOINT
+STATIC_INLINE int copystp(char *dest, const char *src)
 {
-    // Our telemetry on SPCS expects a JSON object per line
-    // The following lines prepare the timestamp string and the JSON object
+    char *d = stpcpy(dest, src);
+    return (int)(d - dest);
+}
+
+// RAI-specific
+STATIC_INLINE void write_to_safe_crash_log(char *buf) JL_NOTSAFEPOINT
+{
+    // Our telemetry on SPCS expects a JSON object per line.
+    // We ignore write failures because there is nothing we can do.
+    int buflen = strlen(buf);
+    assert(buflen <= 4096);
+    char wbuf[4096];
+    bzero(wbuf, 4096);
+    int wlen = 0;
+
+    // JSON preamble
+    wlen += copystp(&wbuf[wlen], "\n{\"level\":\"Error\", \"timestamp\":\"");
+
+    // Timestamp
     struct timeval tv;
     struct tm* tm_info;
-    char timestamp_buffer[50];
-    // Get current time
     gettimeofday(&tv, NULL);
     tm_info = gmtime(&tv.tv_sec);
-    // Format time
-    int offset = strftime(timestamp_buffer, 25, "%Y-%m-%dT%H:%M:%S", tm_info);
-    // Append milliseconds
-    snprintf(timestamp_buffer + offset, 25, ".%03d", tv.tv_usec / 1000);
-    const char *json_preamble_p1 = "\n{\"level\":\"Error\", \"timestamp\":\"";
-    const char *json_preamble_p2 = "\", \"message\": \"";
-    const char *json_postamble = "\"}\n";
-    // Ignore write failures because there is nothing we can do
-    write(jl_sig_fd, json_preamble_p1, strlen(json_preamble_p1));
-    write(jl_sig_fd, timestamp_buffer, strlen(timestamp_buffer));
-    write(jl_sig_fd, json_preamble_p2, strlen(json_preamble_p2));
-    // JSON escape the input string
-    for(size_t i = 0; i < strlen(buf); i += 1) {
+    wlen += strftime(&wbuf[wlen], 42, "%Y-%m-%dT%H:%M:%S", tm_info);
+    sprintf(&wbuf[wlen], ".%03ld", (long)tv.tv_usec / 1000);
+    wlen += 4;
+
+    // JSON preamble to message
+    wlen += copystp(&wbuf[wlen], "\", \"message\": \"");
+
+    // Message
+    for (size_t i = 0; i < buflen; i++) {
```

Comment:
Currently this can over run the buffer, because of the preamble, I think?

Reply:
And the closing chars

Reply:
Improved the buffer math.